# Module 2 Summative Lab

## Import Libraries & Connect to Database

In [1]:
# Import necessary libraries

import sqlite3
import pandas as pd
import matplotlib.pyplot as plt
import json
import requests
import pymongo

In [2]:
# Connect to database
conn = sqlite3.connect('''database.sqlite''')

# Create cursor object
cur = conn.cursor()

## Preview Datasets in Database

The SQL database was downloaded from https://www.kaggle.com/laudanum/footballdelphi. The database includes four tables:
 - Matches
 - Teams
 - Teams_in_Matches
 - Unique_Teams

In [3]:
# Preview Matches dataframe and shape

cur.execute('''SELECT * FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
print(Matches_df.shape)
Matches_df.head()

(24625, 9)


Match_ID Div  Season        Date       HomeTeam        AwayTeam  FTHG  \
0         1  D2    2009  2010-04-04     Oberhausen  Kaiserslautern     2   
1         2  D2    2009  2009-11-01    Munich 1860  Kaiserslautern     0   
2         3  D2    2009  2009-10-04  Frankfurt FSV  Kaiserslautern     1   
3         4  D2    2009  2010-02-21  Frankfurt FSV       Karlsruhe     2   
4         5  D2    2009  2009-12-06          Ahlen       Karlsruhe     1   

   FTAG FTR  
0     1   H  
1     1   A  
2     1   D  
3     1   H  
4     3   A

There are 24,625 rows and 9 columns in the __Matches__ table. The columns in this dataset are described from the source (https://www.kaggle.com/laudanum/footballdelphi) as follows:
 - Match_ID (int): unique ID per match
 - Div (str): identifies the division the match was played in (D1 = Bundesliga, D2 = Bundesliga 2, E0 = English Premier League)
 - Season (int): Season the match took place in (usually covering the period of August till May of the following year)
 - Date (str): Date of the match
 - HomeTeam (str): Name of the home team
 - AwayTeam (str): Name of the away team
 - FTHG (int) (Full Time Home Goals): Number of goals scored by the home team
 - FTAG (int) (Full Time Away Goals): Number of goals scored by the away team
 - FTR (str) (Full Time Result): 3-way result of the match (H = Home Win, D = Draw, A = Away Win)

In [4]:
# Preview Teams dataframe and shape

cur.execute('''SELECT * FROM Teams;''')
Teams_df = pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
print(Teams_df.shape)
Teams_df.head()

(468, 8)


Season       TeamName KaderHome AvgAgeHome ForeignPlayersHome  \
0   2017  Bayern Munich        27         26                 15   
1   2017       Dortmund        33         25                 18   
2   2017     Leverkusen        31         24                 15   
3   2017     RB Leipzig        30         23                 15   
4   2017     Schalke 04        29         24                 17   

  OverallMarketValueHome AvgMarketValueHome StadiumCapacity  
0              597950000           22150000           75000  
1              416730000           12630000           81359  
2              222600000            7180000           30210  
3              180130000            6000000           42959  
4              179550000            6190000           62271

There are 468 rows and 8 columns included in the __Teams__ table. The columns in this dataset are described from the source (https://www.kaggle.com/laudanum/footballdelphi) as follows:
 - Season (str): Football season for which the data is valid
 - TeamName (str): Name of the team the data concerns
 - KaderHome (str): Number of Players in the squad
 - AvgAgeHome (str): Average age of players
 - ForeignPlayersHome (str): Number of foreign players (non-German, non-English respectively) playing for the team
 - OverallMarketValueHome (str): Overall market value of the team pre-season in EUR (based on data from transfermarkt.de)
 - AvgMarketValueHome (str): Average market value (per player) of the team pre-season in EUR (based on data from transfermarkt.de)
 - StadiumCapacity (str): Maximum stadium capacity of the team's home stadium

In [5]:
# Preview Teams_in_Matches dataframe and shape

cur.execute('''SELECT * FROM Teams_in_Matches;''')
Teams_in_Matches_df = pd.DataFrame(cur.fetchall())
Teams_in_Matches_df.columns = [x[0] for x in cur.description]
print(Teams_in_Matches_df.shape)
Teams_in_Matches_df.head()

(49148, 2)


Match_ID  Unique_Team_ID
0         1              26
1         1              46
2         2              26
3         2              42
4         3              26

There are 49,148 rows and 2 columns included in the __Teams_in_Matches__ table. The columns in this dataset are described from the source (https://www.kaggle.com/laudanum/footballdelphi) as follows:
 - Match_ID (int): Unique match ID
 - Unique_Team_ID (int): Unique team ID (This table is used to easily retrieve each match a given team has played in)

In [6]:
# Preview Unique_Teams dataframe and shape

cur.execute('''SELECT * FROM Unique_Teams;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
print(Unique_Teams_df.shape)
Unique_Teams_df.head()

(128, 2)


TeamName  Unique_Team_ID
0  Bayern Munich               1
1       Dortmund               2
2     Leverkusen               3
3     RB Leipzig               4
4     Schalke 04               5

There are 128 rows and 2 columns included in the __Unique_Teams__ table. The columns in this dataset are described from the source (https://www.kaggle.com/laudanum/footballdelphi) as follows:
 - TeamName (str): Name of a team
 - Unique_Team_ID (int): Unique identifier for each team

The lab instructions require that the data used in this project is for the 2011 season only. Below are the matches that are included in the 2011 season from the __Matches__ data table.

In [7]:
# Preview 2011 Season Matches from Matches dataset

cur.execute('''SELECT * 
               FROM Matches
               WHERE Season = 2011;''')
Matches_2011_df =pd.DataFrame(cur.fetchall())
Matches_2011_df.columns = [x[0] for x in cur.description]
print(Matches_2011_df.shape)
Matches_2011_df.head()

(992, 9)


Match_ID Div  Season        Date   HomeTeam       AwayTeam  FTHG  FTAG FTR
0      1092  D1    2011  2012-03-31   Nurnberg  Bayern Munich     0     1   A
1      1093  D1    2011  2011-12-11  Stuttgart  Bayern Munich     1     2   A
2      1094  D1    2011  2011-08-13  Wolfsburg  Bayern Munich     0     1   A
3      1095  D1    2011  2011-11-27      Mainz  Bayern Munich     3     2   H
4      1096  D1    2011  2012-02-18   Freiburg  Bayern Munich     0     0   D

# Calculate Summary Statistics

Only using the data from the downloaded SQL tables, the following must be selected/computed:
 - The name of the team
 - The total number of goals scored by the team during the 2011 season
 - The total number of wins the team earned during the 2011 season
Only the __Matches__ table is required to get the above information. The below SQL query uses aggregations,  case-statements and a self-join to return the required information in a dataframe, __Stats_df__.

In [8]:
# Create a dataframe that shows the number of wins, draws and loses for each team during the 2011 season

cur.execute('''SELECT HomeTeam AS TeamName, (HomeGoalCount + AwayGoalCount) AS TotalGoals, (WinsH+WinsA) AS Wins, (DrawsH+DrawsA) AS Draws, (LosesH+LosesA) AS Loses
FROM (
SELECT HomeTeam, 
SUM(FTHG) AS HomeGoalCount,
SUM(CASE FTR
    WHEN 'H' THEN 1
    WHEN 'D' THEN 0
    WHEN 'A' THEN 0
END) AS WinsH,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 1
    WHEN 'A' THEN 0
END) AS DrawsH,
SUM (CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 0
    WHEN 'A' THEN 1
END) AS LosesH
FROM Matches
WHERE Season = '2011'
GROUP BY HomeTeam) AS t1
JOIN
(SELECT AwayTeam,
SUM(FTAG) As AwayGoalCount,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 0
    WHEN 'A' THEN 1
END) AS WinsA,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 1
    WHEN 'A' THEN 0
END) AS DrawsA,
SUM (CASE FTR
    WHEN 'H' THEN 1
    WHEN 'D' THEN 0
    WHEN 'A' THEN 0
END) AS LosesA
FROM Matches
WHERE Season = '2011'
GROUP BY AwayTeam) AS t2
ON
t1.HomeTeam = t2.AwayTeam;''')
Stats_df = pd.DataFrame(cur.fetchall())
Stats_df.columns = [x[0] for x in cur.description]
Stats_df.head()

TeamName  TotalGoals  Wins  Draws  Loses
0         Aachen          30     6     13     15
1        Arsenal          74    21      7     10
2    Aston Villa          37     7     17     14
3       Augsburg          36     8     14     12
4  Bayern Munich          77    23      4      7

# DarkSky API WeatherGetter

Next, we need to compute each team's win percentage on days where it was raining during games in the 2011 season. The below class will take in a list of dates and get the weather in Berlin on that date. It will also produce each team's percentage of wins that occured in the rain.

In [9]:
class WeatherGetter:
    
    def __init__(self):
        self.path = "/Users/erikadauria/Flatiron/Projects/Soccer_in_the_Rain/.secret/darksky_api.json"
        self.url = 'https://api.darksky.net/forecast/'
        self.berlin_latitude = '52.520008'
        self.berlin_longitude = '13.404954'
        self.exclude = '?exclude=minutely,hourly,currently,flags,alerts'
    
    def get_key(self):
        with open(self.path) as f:
            key = json.load(f)
            self.api_key = key['api_key']
            return self.api_key
    
    def get_weather(self, day):
        api_key = self.get_key()
        date = day+'T12:00:00'
        request_string = '{}{}/{},{},{}{}'.format(self.url, api_key, self.berlin_latitude, self.berlin_longitude, date,self.exclude)
        response = requests.get(request_string)
        self.weather = response.json()['daily']['data'][0]['icon']
        return self.weather
    
    def get_weather_for_all_days(self, dates):
        self.weather_df = pd.DataFrame(dates, columns = ['Date'])
        self.weather_df['Weather'] = ''
        for i in range(len(self.weather_df.Date)):
            self.weather_df['Weather'][i]= self.get_weather(self.weather_df['Date'][i])
        return self.weather_df
    
    def get_rain_win_percentage(self, dataset):
        final_df = pd.merge(dataset, self.weather_df, on='Date')
        final_df = final_df[final_df['Weather'] == 'rain'].reset_index()
        teams = set(final_df['HomeTeam'])
        rain_dict = {}
        for team in teams:
            total = 0
            wins = 0
            for i in range(len(final_df)):
                if (final_df['HomeTeam'][i] == team):
                    total += 1
                    if final_df['FTR'][i] == 'H':
                        wins += 1
                elif final_df['AwayTeam'][i] == team:
                    total +=1
                    if final_df['FTR'][i] == 'A':
                        wins += 1
            win_percent = round((wins/total)*100,2)
            rain_dict.update({team: win_percent})
        rain_win_df = pd.DataFrame.from_dict(rain_dict, orient='index').reset_index()
        rain_win_df.columns = ['TeamName', 'RainWinPercentage']
        return rain_win_df


In [10]:
# Create instance of WeatherGetter class

weather = WeatherGetter()

In [12]:
weather.get_weather('2012-03-31')

'rain'

In [13]:
# Get weather data for each day there was a match during the 2011 season

weather_df = weather.get_weather_for_all_days(Matches_2011_df.Date.unique())
weather_df.head()

KeyError: 'icon'

In [ ]:

rain_win_df = weather.get_rain_win_percentage(Matches_2011_df)
rain_win_df.head()

# Getting the Wins and Loses for Each Team with SQL

In [ ]:
# Create a dataframe that shows the number of wins, draws and loses for each team during the 2011 season

cur.execute('''SELECT HomeTeam AS TeamName, (HomeGoalCount + AwayGoalCount) AS TotalGoals, (WinsH+WinsA) AS Wins, (DrawsH+DrawsA) AS Draws, (LosesH+LosesA) AS Loses
FROM (
SELECT HomeTeam, 
SUM(FTHG) AS HomeGoalCount,
SUM(CASE FTR
    WHEN 'H' THEN 1
    WHEN 'D' THEN 0
    WHEN 'A' THEN 0
END) AS WinsH,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 1
    WHEN 'A' THEN 0
END) AS DrawsH,
SUM (CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 0
    WHEN 'A' THEN 1
END) AS LosesH
FROM Matches
WHERE Season = '2011'
GROUP BY HomeTeam) AS t1
JOIN
(SELECT AwayTeam,
SUM(FTAG) As AwayGoalCount,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 0
    WHEN 'A' THEN 1
END) AS WinsA,
SUM(CASE FTR
    WHEN 'H' THEN 0
    WHEN 'D' THEN 1
    WHEN 'A' THEN 0
END) AS DrawsA,
SUM (CASE FTR
    WHEN 'H' THEN 1
    WHEN 'D' THEN 0
    WHEN 'A' THEN 0
END) AS LosesA
FROM Matches
WHERE Season = '2011'
GROUP BY AwayTeam) AS t2
ON
t1.HomeTeam = t2.AwayTeam;''')
Stats_df = pd.DataFrame(cur.fetchall())
Stats_df.columns = [x[0] for x in cur.description]
Stats_df.head()

In [ ]:
# Merge rain_win_df and Stats_df

final_df = pd.merge(Stats_df, rain_win_df, on='TeamName')
final_df.head()

In [ ]:
import gridfs

In [ ]:
# Function to create histogram

def create_histogram(ind):
    data = final_df.iloc[ind].to_dict()
    x_values = list(data.keys())[2:5]
    y_values = list(data.values())[2:5]
    fig = plt.figure(figsize=(5,5))
    plt.title(list(data.values())[0] + ' 2011 Season Stats')
    plt.bar(x_values, y_values)
    plt.savefig('Histograms/' + list(data.values())[0] + '.png', bbox_inches='tight')
    plt.close(fig)
    

In [ ]:
client=MongoClient()
db=client.mytest
data=open(“image.png”,”rb”)
fs=gridfs.GridFS(db)
thedata=data.read()
stored=fs.put(thedata,filename=”inmongoimage”)

To get it back in the same code you call:
out=fs.get(stored).read()

In [ ]:
create_histogram(0)

# MongoDB

In [ ]:
from PIL import Image
from bson import Binary
import io

#img = Image.open('test.jpg')

#imgByteArr = io.BytesIO()
#img.save(imgByteArr, format='PNG')
#imgByteArr = imgByteArr.getvalue()

In [ ]:
class MongoHandler:
    def __init__(self):
        self.myclient = pymongo.MongoClient()
        self.database = self.myclient['soccer_stats_database']
        self.collection = self.database['soccer_stats_collection']
        
    def reformat_png(self, png_file):
        data=open(png_file, "rb")        
        fs=gridfs.GridFS(self.database)
        thedata=data.read()
        stored=fs.put(thedata, filename="mongoimage")
        return fs.get(stored).read()
    
    def format_data(self, ind):
        self.stats_dict = {
            "team" : final_df.TeamName[i],
            "histogram" : self.reformat_png('Histograms/' + final_df.TeamName[i] + '.png')  
        }
        return self.stats_dict
    
    def insert_record(self, ind):
        #data = soccer_Team_Data_in_dict_Form #{'name': 'John Doe', 'address': '123 elm street', 'age': 28}
        stats_dict = self.format_data(ind)
        return self.collection.insert_one(stats_dict)
    
    def view_mongo_collection(self):
        query = self.collection.find({})
        for x in query:
            print(x)
    
    def delete_collection(self):
        self.collection.delete_many({})

In [ ]:
mongo = MongoHandler()

In [ ]:
for i in range(3):
    create_histogram(i)
    
    mongo.insert_record(i)

In [ ]:
#for i in range(len(final_df)):
for i in range(3):
    create_histogram(i)
    stats_dict = {
        "team" : final_df.TeamName[i],
        "total_goals" : float(final_df.TotalGoals[i]),
        "total_wins" : float(final_df.Wins[i]),
        "histogram" : Image.open('Histograms/' + final_df.TeamName[i] + '.png').show(),
        "rain_win_percentage" : final_df.RainWinPercentage[i]
    }
    mongo.insert_record(stats_dict)

In [ ]:
mongo.view_mongo_collection()

In [ ]:
mongo.delete_collection()

In [ ]:
mongo.view_mongo_collection()